financial-templates-lib/price-feed/utils.js

```diff
@@ -0,0 +1,130 @@
+const lodash = require("lodash");
+const assert = require("assert");
+
+// Downloads blocks and caches them for certain time into the past.
+// Allows some in memory searches to go from timestamp to block number.
+exports.BlockHistory = (web3, blocks = []) => {
+  assert(web3, "requires web3");
+
+  // Check if we have downloaded a block by number
+  function has(number) {
+    return blocks.find(block => block.number === number);
+  }
+
+  // Used internally, but will insert a block into cache sorted by timestamp
+  function insert(block) {
+    const index = lodash.sortedIndexBy(blocks, block, "timestamp");
+    blocks.splice(index, 0, block);
+    return blocks;
+  }
+
+  // Get the block with the closest match to timestamp
+  function getClosestTime(timestamp) {
+    // const index = lodash.sortedIndexBy(blocks,{timestamp},'timestamp')
+    const index = lodash.sortedIndexBy(blocks, { timestamp }, x => {
+      return x.timestamp;
+    });
+    return blocks[index];
+  }
+
+  // Call to update cache with blocks between sometime in the past from a block number
+  // Block number is optional, if not specified will default to latest block number.
+  async function fetchBetween(endTime, blockNumber) {
+    blockNumber = blockNumber || (await web3.eth.getBlockNumber());
+    if (has(blockNumber)) return fetchBetween(endTime, blockNumber - 1);
+    const block = await web3.eth.getBlock(blockNumber);
+    assert(block, "Provider returned empty block");
+    if (block.timestamp < endTime) return blocks;
+    insert(block);
+    return fetchBetween(endTime, blockNumber - 1);
+  }
+
+  // Main call to update cache, will take care of fetching all blocks, caching and pruning cache.
+  async function update(age, now, startBlock) {
+    assert(age, "requires age in seconds");
+    assert(now, "requires current time");
+    const endTime = now - age;
+    blocks = pruneByTimestamp(age);
+    const result = fetchBetween(endTime, startBlock);
+    return result;
+  }
+
+  // Removes blocks from cache which are older than age
+  function pruneByTimestamp(age) {
+    return blocks.filter(block => block.timestamp > seconds);
+  }
+
+  // Return all blocks in cache
+  function listBlocks() {
+    return blocks;
+  }
+  return {
+    has,
+    insert,
```

Comment:
Bump on this comment. Are some of these methods meant to be internal? I think exporting fewer functions and leaving the ones that are only expected to be used internally as private may simplify the interface a little. It could also help to sort their implementations this way as well.

Reply:
the reason i do expose private functions is primarily for testing. theres no way to reach into the closure if the functions are not exposed. to avoid incorrect usage of the class it would need to be documented. I havent really figured out a good way to prevent functions from exposed to user while also being able to test them.

Reply:
Yep, makes sense. I think giving yourself more confidence in the code is more important than worrying about "not testing implementation details". I think a comment and sorting the file according to public and private would probably help make the distinction clear.